src/azure-cli-core/azure/cli/core/_profile.py

```diff
@@ -117,6 +115,8 @@ def login(self,
               password,
               is_service_principal,
               tenant,
+              scopes=None,
+              client_id=None,
```

Comment:
why need specify scopes when login? I assume you only need to use it when get_token for the service API call

Reply:
This is due to the fact that a 3rd party application does not always have access to the common management scope/resource (e.g. to the https://management.core.windows.net/). 

For example, this is the error message I receive if I do not specify the proper scope in the login call:

Authentication failed: ["AADSTS650057: Invalid resource. The client has requested access to a resource which is not listed in the requested permissions in the client's application registration. Client app ID: 0b1c8f7e-28d2-4378-97e2-7d7d63f7c87f(Azure Sphere Utility). Resource value from request: https://management.core.windows.net/. Resource app ID: 797f4846-ba00-4fd7-ba43-dac1f8f63013. List of valid resources from app registration: f0277259-56be-4121-a579-5a57fcd1dbac, 00000002-0000-0000-c000-000000000000.\r\nTrace ID: 820a9c22-4c5b-44c5-82ca-1e5675584800\r\nCorrelation ID: fbfb8bfc-1421-4241-820a-e05940c18b01\r\nTimestamp: 2020-07-02 12:12:54Z"]

By adding the `scopes` the parameter to the `login` function, the "log in" call means, that the user logs in into an AAD `tenant` from an app specified by the `client_id` to access a resource specified by the `scopes`.

What do you think?

Reply:
Reading your other comment about `cli_ctx`, just had a thought about this. Would it be a feasible option to overwrite the `cli_ctx.cloud.endpoints.active_directory_resource_id` by the `login` command? If yet, would it be an option to introduce a `cli_ctx.cloud.endpoints.active_directory_scopes` member to be compatible with MSAL?

Reply:
@qianwens, according to [authorization code flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow#request-an-authorization-code), the initial request also requires a `scope` to which the `client_id` may not have access:

```
https://login.microsoftonline.com/{tenant}/oauth2/v2.0/authorize?
client_id=6731de76-14a6-49ae-97bc-6eba6914391e
&response_type=code
&redirect_uri=http%3A%2F%2Flocalhost%2Fmyapp%2F
&response_mode=query
&scope=openid%20offline_access%20https%3A%2F%2Fgraph.microsoft.com%2Fmail.read
&state=12345
```

That's why we need to make it customizable.

Reply:
Azure Identity has some built-in default scopes:

https://github.com/Azure/azure-sdk-for-python/blob/fdf11b302b20821e94eae61a2eabd8eb1d776162/sdk/identity/azure-identity/azure/identity/_internal/msal_credentials.py#L41-L46

```py
_DEFAULT_AUTHENTICATE_SCOPES = {
    "https://" + KnownAuthorities.AZURE_CHINA: ("https://management.core.chinacloudapi.cn//.default",),
    "https://" + KnownAuthorities.AZURE_GERMANY: ("https://management.core.cloudapi.de//.default",),
    "https://" + KnownAuthorities.AZURE_GOVERNMENT: ("https://management.core.usgovcloudapi.net//.default",),
    "https://" + KnownAuthorities.AZURE_PUBLIC_CLOUD: ("https://management.core.windows.net//.default",),
}
```

Reply:
Azure Identity has some built-in default scopes:

https://github.com/Azure/azure-sdk-for-python/blob/fdf11b302b20821e94eae61a2eabd8eb1d776162/sdk/identity/azure-identity/azure/identity/_internal/msal_credentials.py#L41-L46

```py
_DEFAULT_AUTHENTICATE_SCOPES = {
    "https://" + KnownAuthorities.AZURE_CHINA: ("https://management.core.chinacloudapi.cn//.default",),
    "https://" + KnownAuthorities.AZURE_GERMANY: ("https://management.core.cloudapi.de//.default",),
    "https://" + KnownAuthorities.AZURE_GOVERNMENT: ("https://management.core.usgovcloudapi.net//.default",),
    "https://" + KnownAuthorities.AZURE_PUBLIC_CLOUD: ("https://management.core.windows.net//.default",),
}
```

Reply:
This is correct. It picks up the one which belongs to the passed authority.

Reply:
Possibly we can move the `client_id` to `__init__`?

Reply:
I have moved both `client_id` and `scopes` into the constructor.